src/function/aggregate/sorted_aggregate_function.cpp

```diff
@@ -359,44 +377,97 @@ struct SortedAggregateFunction {
+
+		// First pass: Sort all the input payloads at once on (state_idx ASC, orders)
+		vector<BoundOrderByNode> orders;
+		orders.emplace_back(BoundOrderByNode(OrderType::ASCENDING, OrderByNullType::NULLS_FIRST,
+		                                     make_uniq<BoundConstantExpression>(Value::USMALLINT(0))));
+		for (const auto &order : order_bind->orders) {
+			orders.emplace_back(order.Copy());
+		}
+
+		GlobalSortState global_sort(buffer_manager, orders, payload_layout);
+		LocalSortState local_sort;
+		local_sort.Initialize(global_sort, global_sort.buffer_manager);
+
+		DataChunk prefixed;
+		prefixed.Initialize(Allocator::DefaultAllocator(), global_sort.sort_layout.logical_types);
+
+		vector<idx_t> state_counts(count, 0);
 		for (idx_t i = 0; i < count; ++i) {
-			initialize(agg_state.data());
 			auto state = sdata[i];
+			prefixed.Reset();
+			prefixed.data[0].Reference(Value::USMALLINT(i));
 
-			// Apply the sort before delegating the chunks
-			auto global_sort = make_uniq<GlobalSortState>(buffer_manager, orders, payload_layout);
-			LocalSortState local_sort;
-			local_sort.Initialize(*global_sort, global_sort->buffer_manager);
-			state->Finalize(*order_bind, local_sort);
-			global_sort->AddLocalState(local_sort);
-
-			if (!global_sort->sorted_blocks.empty()) {
-				global_sort->PrepareMergePhase();
-				while (global_sort->sorted_blocks.size() > 1) {
-					global_sort->InitializeMergeRound();
-					MergeSorter merge_sorter(*global_sort, global_sort->buffer_manager);
-					merge_sorter.PerformInMergeRound();
-					global_sort->CompleteMergeRound(false);
-				}
+			state_counts[i] = state->Finalize(*order_bind, prefixed, local_sort);
+		}
+		global_sort.AddLocalState(local_sort);
+
+		//	Second pass: scan the sorted data until we are done and pass slices to be aggregated.
+		idx_t i = 0;
+		if (!global_sort.sorted_blocks.empty()) {
+			//	Sort all the data
+			global_sort.PrepareMergePhase();
+			while (global_sort.sorted_blocks.size() > 1) {
+				global_sort.InitializeMergeRound();
+				MergeSorter merge_sorter(global_sort, global_sort.buffer_manager);
+				merge_sorter.PerformInMergeRound();
+				global_sort.CompleteMergeRound(false);
+			}
 
-				PayloadScanner scanner(*global_sort);
-				for (;;) {
-					chunk.Reset();
-					scanner.Scan(chunk);
-					if (chunk.size() == 0) {
-						break;
+			PayloadScanner scanner(global_sort);
+			initialize(agg_state.data());
+			while (scanner.Remaining()) {
+				chunk.Reset();
+				scanner.Scan(chunk);
+				idx_t consumed = 0;
+
+				// Distribute the rows to the aggregates
+				while (consumed < chunk.size()) {
+					//	Find the next aggregate that needs data
+					for (; !state_counts[i]; ++i) {
+						// Finalize a single value at the next offset
+						agg_state_vec.SetVectorType(states.GetVectorType());
+						finalize(agg_state_vec, aggr_bind_info, result, 1, i + offset);
+						if (destructor) {
+							destructor(agg_state_vec, 1);
```

Comment:
Note: this will conflict with the addition of the `aggr_bind_info` in the destructor in #6995 (but easy to resolve)